src/core/read_halos-velociraptor.c

```diff
@@ -140,180 +118,153 @@ void read_trees__velociraptor(int snapshot,
+
   double mass_unit_to_internal = 1.0;
   double scale_factor = -999.;
 
   *n_halos = 0;
   *n_fof_groups = 0;
 
-  int buffer_size = 10000; // NOTE: Arbitrary. Should be a multiple of the chunk size of arrays in file ideally?
-  tree_entry_t* tree_entries = malloc(sizeof(tree_entry_t) * buffer_size);
-
-  if (run_globals.mpi_rank == 0) {
-    char fname[STRLEN * 2 + 8];
-    switch (run_globals.params.TreesID) {
-      case VELOCIRAPTOR_TREES:
-        sprintf(fname, "%s/trees/%s", run_globals.params.SimulationDir, run_globals.params.CatalogFilePrefix);
-        break;
-      case VELOCIRAPTOR_TREES_AUG:
-        sprintf(fname, "%s/augmented_trees/%s", run_globals.params.SimulationDir, run_globals.params.CatalogFilePrefix);
-        break;
-      default:
-        mlog_error("Unrecognised input trees identifier (TreesID).");
-        break;
-    }
+  char fname[STRLEN * 2 + 8];
+  switch (run_globals.params.TreesID) {
+    case VELOCIRAPTOR_TREES:
+      sprintf(fname, "%s/trees/%s", run_globals.params.SimulationDir, run_globals.params.CatalogFilePrefix);
+      break;
+    case VELOCIRAPTOR_TREES_AUG:
+      sprintf(fname, "%s/augmented_trees/%s", run_globals.params.SimulationDir, run_globals.params.CatalogFilePrefix);
+      break;
+    default:
+      mlog_error("Unrecognised input trees identifier (TreesID).");
+      break;
+  }
     
-    fd = H5Fopen(fname, H5F_ACC_RDONLY, H5P_DEFAULT);
-    if (fd < 0) {
-      mlog("Failed to open file %s", MLOG_MESG, fname);
-      ABORT(EXIT_FAILURE);
-    }
-
-    char snap_group_name[9];
-    sprintf(snap_group_name, "Snap_%03d", snapshot);
-    snap_group = H5Gopen(fd, snap_group_name, H5P_DEFAULT);
-
-    H5LTget_attribute_int(fd, snap_group_name, "NHalos", &n_tree_entries);
-
-    property_buffer = malloc(buffer_size * sizeof(long));
-
-    // check the units
-    H5LTget_attribute_double(fd, "Header/Units", "Mass_unit_to_solarmass", &mass_unit_to_internal);
-    mass_unit_to_internal /= 1.0e10;
-    H5LTget_attribute_double(fd, snap_group_name, "scalefactor", &scale_factor);
+  hid_t fd = H5Fopen(fname, H5F_ACC_RDONLY, plist_id);
+  if (fd < 0) {
+    mlog("Failed to open file %s", MLOG_MESG, fname);
+    ABORT(EXIT_FAILURE);
   }
-
-  MPI_Bcast(&n_tree_entries, 1, MPI_INT, 0, run_globals.mpi_comm);
+  H5Pclose(plist_id);
+
+  int mpi_size = run_globals.mpi_size;
+  int mpi_rank = run_globals.mpi_rank;
+
+  char snap_group_name[9];
+  sprintf(snap_group_name, "Snap_%03d", snapshot);
+  hid_t snap_group = H5Gopen(fd, snap_group_name, H5P_DEFAULT);
+
+  H5LTget_attribute_int(fd, snap_group_name, "NHalos", &n_tree_entries);
+
+  // check the units
+  H5LTget_attribute_double(fd, "Header/Units", "Mass_unit_to_solarmass", &mass_unit_to_internal);
+  mass_unit_to_internal /= 1.0e10;
+  H5LTget_attribute_double(fd, snap_group_name, "scalefactor", &scale_factor);
+
+  int buffer_size = (n_tree_entries > 100000) ? n_tree_entries / 10 : 10000;
```

Comment:
Just to make sure... This is because we chose to optimize it for the chunk size = 10000 right? 
Would it make sense to put a comment here?

Reply:
done